routing/routing.go

```diff
@@ -160,11 +166,44 @@ func (m *LBMetrics) DecInflightRequest() {
 	atomic.AddInt64(&m.inflightRequests, -1)
 }
 
-// GetInflightRequests decrements the number of outstanding requests from the proxy to a given backend.
+// GetInflightRequests loads the number of outstanding requests from the proxy to a given backend.
 func (m *LBMetrics) GetInflightRequests() int {
 	return int(atomic.LoadInt64(&m.inflightRequests))
 }
 
+// AddFailedRequest increments the counters related to failed requests and total requests from the proxy to a given backend.
+func (m *LBMetrics) AddFailedRequest() {
+	m.m.Lock()
+	defer m.m.Unlock()
+
+	m.failedRequests++
+	m.totalRequests++
```

Comment:
why not 
```
atomic.AddInt64(&m.failedRequests, 1)
atomic.AddInt64(&m.totalRequests, 1)
```

Reply:
Because in this section of code `m.m` is locked, so we can rely on no other goroutine using those fields.
So both usual and atomic instructions are correct.
However, atomic ones are slower, because they add `mfence` instructions, like here https://stackoverflow.com/questions/51846894/what-is-the-performance-of-stdatomic-vs-non-atomic-variables

Reply:
I'm not sure if this post is the same case we are seeing here. They are explicitly comparing a normal operation vs an atomic one, but the performance constraints also involve the mutex lock. Atomic instructions are busy waiting, mutex free up CPU for other threads. I can't tell one or the other is more efficient for our use case without benchmarks, but our use case seems different from the question in your link. Or I might be imagining things, either way its a valid discussion. As food for thought: https://stackoverflow.com/questions/15056237/which-is-more-efficient-basic-mutex-lock-or-atomic-integer

Reply:
I swapped the lines, because you want to add total before failed, such that you do not count failed more than total (could lead to failed is bigger than total).
In general I am sure that lock will be slower as @lucastt summarized. If you like to check it, create a micro benchmark. Do not trust random posts by people, even if this would be super-hero-popular-10x-coder, it can be wrong.

Reply:
Make sure you increase first total requests and after that the failed ones.